src/core/lib/security/credentials/google_default/google_default_credentials.cc

```diff
@@ -273,59 +274,82 @@ static grpc_error* create_default_creds_from_path(
   return error;
 }
 
-grpc_channel_credentials* grpc_google_default_credentials_create() {
-  grpc_channel_credentials* result = nullptr;
+static void update_tenancy() {
+  gpr_once_init(&g_once, init_default_credentials);
+  gpr_mu_lock(&g_state_mu);
+
+  /* Try a platform-provided hint for GCE. */
+  if (!g_metadata_server_available) {
+    g_metadata_server_available = g_gce_tenancy_checker();
+  }
+  /* TODO: Add a platform-provided hint for GAE. */
+
+  /* Do a network test for metadata server. */
+  if (!g_metadata_server_available) {
+    g_metadata_server_available = is_metadata_server_reachable();
+  }
+  gpr_mu_unlock(&g_state_mu);
+}
+
+static bool metadata_server_available() {
+  bool available = false;
+  gpr_mu_lock(&g_state_mu);
+  available = static_cast<bool>(g_metadata_server_available);
+  gpr_mu_unlock(&g_state_mu);
+  return available;
+}
+
+static grpc_core::RefCountedPtr<grpc_call_credentials> make_default_call_creds(
+    grpc_error** error) {
   grpc_core::RefCountedPtr<grpc_call_credentials> call_creds;
-  grpc_error* error = GRPC_ERROR_CREATE_FROM_STATIC_STRING(
-      "Failed to create Google credentials");
   grpc_error* err;
-  grpc_core::ExecCtx exec_ctx;
 
-  GRPC_API_TRACE("grpc_google_default_credentials_create(void)", 0, ());
-
-  gpr_once_init(&g_once, init_default_credentials);
+  update_tenancy();
```

Comment:
Can we move this to line 323? The tenancy check does not seem necessary if the call credentials is not read from the metadata server. By doing that, it will be also consistent with the old code path.

Reply:
Implementation fixed and tests modified to encode this expectation.